### PR TITLE
sanitize filename before creating it in the blobs dir

### LINKED ISSUE #2248: Can't share images with "invalid" characters in name
- Delta Chat version: Current master
- Steps to reproduce the problem:
  - Create an image with an "invalid" character in the file name like `/` or `\`. (For some reason, my built-in camera app sometimes creates names like `IMG_3874\.jpg`.)
  - Click "share", select DC, and select a chat. Expected: The image is set as a draft. Actual: It's not.

Ping @r10s can you check whether there is a similar issue in iOS, and if not how you avoid it?

**Reason**
There's just never any code that would remove the invalid character. The Android code directly creates a file in the blob dir (`handleSharing()` -> `createMessage()` -> `getRealPathFromUri()` -> `getBlobdirFile()`) and sets it on the message with `set_file()`, which is a simple setter. Then when it calls `set_draft()`, the core errors `failed to set draft: Blob has a badname /data/.../IMG_394\.jpg`.

**Proposed solution**

Expose `BlobObject::create()` via the FFI and let Android use it instead of `getBlobdirFile()`. Ping @link2xt @flub as this would be a Core change.

**Alternative solutions**

- In DC-Android: Directly fix `getBlobdirFile()`, implementing a logic similar to what `BlobObject::create()` does.
- In DC-Core: Let `set_draft()` or `set_file()` repair the filename, copying the file (or moving if we are sure it's not needed under the old name)

### PR DESCRIPTION
**Pull request by adbenitez**
close #2248

### CODE REVIEW

File: src/org/thoughtcrime/securesms/util/FileUtils.java [hunk 1]
```diff
@@ -1,9 +1,49 @@
...
+    }
+    return res.toString();
+  }
+
+  private static boolean isValidFilenameChar(char c) {
+    if ((0x00 <= c && c <= 0x1f)) {
+      return false;
+    }
+    switch (c) {
+      case '"':
```
Comment (adbenitez) [inline src/org/thoughtcrime/securesms/util/FileUtils.java:33]: maybe this is too strict? (I took it from the FAT filesystem) feel free to tweak and remove the ones you think is fine to allow

File: src/org/thoughtcrime/securesms/util/FileUtils.java [hunk 2]
```diff
@@ -1,9 +1,49 @@
...
+        res.append(c);
+      } else {
+        res.append('_');
+      }
+    }
+    return res.toString();
+  }
+
+  private static boolean isValidFilenameChar(char c) {
+    if ((0x00 <= c && c <= 0x1f)) {
```
Comment (r10s) [inline src/org/thoughtcrime/securesms/util/FileUtils.java:29]: why not just use

```suggestion
    if (c <= 0x1f) {
```

anyways, this check already covers the 'case '\0'` below

File: src/org/thoughtcrime/securesms/util/FileUtils.java [hunk 3]
```diff
@@ -1,9 +1,49 @@
...
+    switch (c) {
+      case '"':
+      case '*':
+      case '/':
+      case ':':
+      case '<':
+      case '>':
+      case '?':
+      case '\\':
+      case '\0':
```
Comment (r10s) [inline src/org/thoughtcrime/securesms/util/FileUtils.java:41]: `\0` is already covered by the range check above

File: src/org/thoughtcrime/securesms/util/FileUtils.java (hunk 1 above)
Comment (r10s) [reply src/org/thoughtcrime/securesms/util/FileUtils.java:33]: i think, it is okay. we're just using the sanitised name and not failing

**Review by r10s — APPROVED** [review]
some minors

File: src/org/thoughtcrime/securesms/util/FileUtils.java (hunk 3 above)
Comment (adbenitez) [reply src/org/thoughtcrime/securesms/util/FileUtils.java]: fixed, danke!

**Pull request merged**